pkg/hbone/dialer.go

```diff
@@ -0,0 +1,173 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package hbone
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"golang.org/x/net/http2"
+	"golang.org/x/net/proxy"
+
+	"istio.io/istio/security/pkg/pki/util"
+	istiolog "istio.io/pkg/log"
+)
+
+var log = istiolog.RegisterScope("hbone", "", 0)
+
+type Config struct {
+	ProxyAddress string
+	Headers      http.Header
+	TLS          *tls.Config
+}
+
+type Dialer interface {
+	proxy.Dialer
+	proxy.ContextDialer
+}
+
+func NewDialer(cfg Config) Dialer {
+	return &dialer{cfg: cfg}
+}
+
+type dialer struct {
+	cfg Config
+}
+
+func (d dialer) DialContext(ctx context.Context, network, address string) (net.Conn, error) {
+	if network != "tcp" {
+		return net.Dial(network, address)
+	}
+	// TODO: use context
+	c, s := net.Pipe()
+	go func() {
+		err := proxyTo(s, d.cfg, address)
+		log.Infof("Tunnel complete: %v", err)
+	}()
+	return c, nil
```

Comment:
I think we need to work on this - see my previous implementation, the HTTP RT should probably be done part of Dial, with proxy remaining in background. So we can return errors properly ( similar with how normal Dial works - i.e. a real connection/stream will be established by dial)

Reply:
Agreed